models/migrations/v115.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"crypto/md5"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+
+	"xorm.io/xorm"
+)
+
+func renameExistingUserAvatarName(x *xorm.Engine) error {
+	sess := x.NewSession()
+	defer sess.Close()
+
+	type User struct {
+		ID        int64  `xorm:"pk autoincr"`
+		LowerName string `xorm:"UNIQUE NOT NULL"`
+		Avatar    string
+	}
+	deleteList := make(map[string]struct{})
+	start := 0
+	for {
+		if err := sess.Begin(); err != nil {
+			return fmt.Errorf("session.Begin: %v", err)
+		}
+		users := make([]*User, 0, 50)
+		if err := sess.Table("user").Asc("id").Limit(50, start).Find(&users); err != nil {
+			return fmt.Errorf("select users from id [%d]: %v", start, err)
+		}
+		if len(users) == 0 {
+			_ = sess.Rollback()
+			break
+		}
+
+		log.Info("select users [%d - %d]", start, start+len(users))
+		start += 50
+
+		for _, user := range users {
+			oldAvatar := user.Avatar
+
+			if _, err := os.Stat(filepath.Join(setting.AvatarUploadPath, oldAvatar)); err != nil {
+				log.Warn("[user: %s] os.Stat: %v", user.LowerName, err)
+				// avatar doesn't exist in the storage
+				// no need to move avatar and update database
+				// we can just skip this
+				continue
+			}
+
+			newAvatar, err := copyOldAvatarToNewLocation(user.ID, oldAvatar)
+			if err != nil {
+				_ = sess.Rollback()
+				return fmt.Errorf("[user: %s] %v", user.LowerName, err)
+			}
+
+			user.Avatar = newAvatar
+			if _, err := sess.ID(user.ID).Cols("avatar").Update(user); err != nil {
+				_ = sess.Rollback()
+				return fmt.Errorf("[user: %s] user table update: %v", user.LowerName, err)
+			}
+
+			deleteList[filepath.Join(setting.AvatarUploadPath, oldAvatar)] = struct{}{}
```

Comment:
What if there is a failure half-way through this migration and so you rerun it?

It looks like in that case you will delete all of the previously migrated newAvatars?

Shouldn't you be testing that the oldAvatar != newAvatar?

Reply:
https://github.com/go-gitea/gitea/pull/8547/files#r361710833

https://github.com/go-gitea/gitea/pull/8547/files#r361710313

should fix this @zeripath ?

Reply:
Thanks. when I wrote this, every time the newAvatar would become different. so back then it wasn't necessary. But now it's not good. I will update as soon as possible. 
Thanks for your suggestion.

Reply:
yeah those would fix it.

Reply:
NP I'll do it for you.